### PR TITLE
Create AttachmentHolder map lazily

### PR DESCRIPTION
**Pull request by embeddedt**
The backing `IdentityHashMap` here can allocate several hundred bytes worth of heap space, which multiplies to a huge amount if there are many `AttachmentHolder` objects in memory. An example is hundreds of thousands of `ItemStack` instances stored by recipes.

To avoid this, we should not create the map until an attachment is actually registered or deserialized.

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java]: That's not correct. One map may have the default instance of an attachment and should therefore be compatible with a `null` map. Maybe store each map or else `Map.of()` if null in local fields.

**Review by Technici4n — DISMISSED** [review]

**Review by marchermans — CHANGES_REQUESTED** [review]
In practice I am okey with this optimization.

But after thinking it over I would like to see some numbers on this improvement. In the past we had a sufficiently large amount of capabilities (in the old system) attached to ItemStacks, and I worry that this optimization may promise more then it delivers.

There are in practice two concepts that I would like to see addressed:
1) Due to the implementation here, the CPU performance of `getData` and `setData` is an order of magnitude worse. This is because the method `getAttachmentMap` is a virtual none private method. As such a method table lookup will occur on every call to either of those methods, regardless of the map already existing or not. The old call used a field access, which is obviously way way faster. This can already be improved by making the method `private final` which would give the JVM the option to inline if the callpath became hot.
2) An alternative idea here, vs option 1) is to use a forced empty IdentityHashMap. The constructor for IHM takes an initial size, which would shrink the inner array down to 0, if 0 is passed in. Severly reducing its memory overhead, and with the field now being none null anyway, we can remove the method call and leave the potentially hot code paths with the old much faster code.

Now on to something else here: For such PRs I would like to see at least some tests, and especially numbers. I know this is hard to test. But for PRs that make claims of memory savings, I would like to see test mods included which would make verification easy. An example here would be a test mod that adds a capability to a % of itemstacks created, then a maintainer can pull a VisualVM profile to do debugging on at least a vanilla world.

Secondly I would like some form of rudimentary CPU profile, if option 2) is not taken. The code paths for getData can become sufficiently hot that a virtual method call and the extra overhead of the null-check might impact performance more then the memory overhead is worth it, depending on the results on the memory test.

Let me make this clear: I agree with this PR, just think there might be other approaches that we can research, and I would like to see some numbers before I think this is ready for merging.

**Review by marchermans — DISMISSED** [review]
After internal discussions:

We think this can be merged.
Please make the `getAttachmentMap` method final, and if this ever becomes a CPU bottle neck we can iterate and further address it.

File: src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java
Comment (shartte) [inline src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java]: This should probably be final like the other methods.

**Review by Technici4n — DISMISSED** [review]

**Review by Technici4n — APPROVED** [review]

**Pull request merged**
